### DISCUSSION TITLE
MIKE1D SDK stepping through simulation and setting new values

### DISCUSSION THREAD

**Discussion by lizaclark** [DHI/Mike1D-SDK-examples#18] [Q&A] [answered]
I am part of a group that has been using MIKE-URBAN 2020 with MOUSE to run a rather large, complex model with user-written controls. We are trying to migrate to MIKE+ 2021 with MIKE-1D. I’ve been attempting to convert our user-written controls from a dll that interacted with MOUSE_hd.dll to a python script that interacts with the various MIKE1D dlls. I am using the controller factory approach detailed in the MIKE Engine SDK MIKE 1D API section of https://docs.mikepoweredbydhi.com/engine_libraries/mike1d/mike1d_api/, more specifically corresponding to https://github.com/DHI/mike1d-sdk-examples/blob/master/DHI.Mike1D.Examples/RunExamples.cs with `controller.EngineNet.PerformTimeStep()` or`controller.RunUntil()` to advance.
 
With MOUSE_hd.dll, we used `DLL_GetNodeLevelPChar`, `DLL_SetNodeLevelPChar`, `DLL_GetGatePositionPChar`, `DLL_SetGatePositionPChar`, `DLL_GetPumpDischargePChar`, and `DLL_SetPumpDischargePChar` to get and set model values in between time steps.
The example of how to use the proxy system to get and set values for node values is pretty clear. I can also GET gate levels and pump discharge using ISensor or IPump directly. What I do not seem to be able to do is *SET* gate levels and pump discharge.
For nodes, the ValueGetter() and ValueSetter() are assigned via the proxy system (the example from the API website below):
```
// Extract a getter, and store that for later use. This should only be done once,
// and the getter should be reused. This should NOT be done everytime a value is required.
 
DDoubleGetter getter;
 
// Proxy system helper class.
var proxyUtil = new ProxyUtil(engineNet);
 
// Get proxy for some node
IDoubleProxy nodeProxy = proxyUtil.GetProxy(engineNet.Nodes[0]);
 
// Get index of water level quantity
quantityIndex = nodeProxy.Offers().IndexOf(Quantity.Create(PredefinedQuantity.WaterLevel))
 
// Getter for water level quantity
if (quantityIndex >= 0)
{
    getter = nodeProxy.ValueGetter(quantityIndex);
}
 
// To get the value.
double value = getter()
``` 
`engineNet` does not include structures (pumps, orifices) or sensors. Is it possible to get/set these via the proxy system?

Does any one have an example (with a similar level of simplicity) of how that would be done? If that's not the right approach, is there an example (with a similar level of simplicity) of the appropriate method for doing this? I think the closest example in this repo to what I'm looking for is the [HonmaCoefficients.cs](https://github.com/DHI/Mike1D-SDK-examples/blob/master/DHI.Mike1D.Examples/Scripts/HonmaWeirCoefficient.cs), but I think it is setting a property of the weir rather than a state or flux variable value. When I set `IOrifice.GateLevel` , the value doesn't update like it does for `HonmaWeir.WeirCoefficient`.

Thanks in advance for any thoughts on this topic.

**Comment by JesperGr** [comment, marked as answer]
It is not possible to set structure values through the proxy system. That only goes for the HD part.

For structures it is necessary to find the specific structure, looping like in [StructureScripts.cs](https://github.com/DHI/Mike1D-SDK-examples/blob/master/DHI.Mike1D.Examples/Scripts/StructureScripts.cs) and searching 
```
      foreach (IStructure structure in mike1DData.Network.StructureCollection.Structures)
      {
        if (structure.ID == "someId")
        {
          [Do stuff]
```
Then for each structure, the approach for controlling the structure is different. 

**Pumps:**
The easiest is to control pump with "constant" flow, by modifying the constant flow at regular intervals:
```
      IStructure structure = [...];
      IPump pump = structure as IPump;
      // Every timestep, or when necessary to change:
      pump.ConstantFlow = 5;
```
If you want the pump to take acceleration and deceleration speeds into account, things becomes slighly more complicated:
```
      IStructure structure = [...];
      Pump pump = structure as Pump;
      // Initially:
      pump.DischargeControlActive = true;
      pump.StartStopLevelControlActive = false;
      // Every timestep, or when necessary to change:
      pump.ControlSetDischarge(5, -1);
```
The pump must not be controlled by other RTC rules, because they will have precedense.

**Weir:**
For weirs, then a weir is unfortunately not just a weir but can be different things. If it ends up as an `OverFlowFormula`. 
```
      OverFlowFormula weir = structure as OverFlowFormula;
      // Every timestep, or when necessary to change:
      weir.CurrentCrestLevel = 5;
```
Again, if taking gate speed into account
```
      OverFlowFormula weir = structure as OverFlowFormula;
      // Every timestep, or when necessary to change:
      weir.ControlSetCrestLevel(5, -1);
```
And it works similarly for an `Orifice`
```
        Orifice weir = structure as Orifice;
      // Every timestep, or when necessary to change:
        weir.GateLevel = 5;
```
and if taking gate speed into account
```
        Orifice weir = structure as Orifice;
      // Every timestep, or when necessary to change:
        weir.ControlSetGateLevel(5, -1);
```
Remember that build in RTC rules must be disabled, otherwise they will take over.
